### ISSUE TITLE
Greader API: FAIL get HTTP Authorization header

### ISSUE THREAD

**Issue by ashwinvis** [YunoHost-Apps/freshrss_ynh#80]
## Problem
I cannot get the greader API working and Feedreader on Linux also fails to connect.

![Screenshot_2019-09-19 FreshRSS API endpoints](https://user-images.githubusercontent.com/9155111/65222955-71d6f680-dac0-11e9-8b00-5c5eb0b49ef9.png)

## Environment

Hardware: Raspberry Pi
FreshRSS Version: 1.14.2
uname -a:

    Linux <hostname:> 4.19.58-v7+ #1245 SMP Fri Jul 12 17:25:51 BST 2019 armv7l GNU/Linux

## Tried so far

Installed missing dependencies, setting up API password, correcting base_url etc one by one, and restarting nginx with no luck:. The following are related threads:

* #69
* https://freshrss.github.io/FreshRSS/en/admins/02_Installation.html
* https://freshrss.github.io/FreshRSS/en/users/06_Mobile_access.html
* https://forum.yunohost.org/t/freshrss-webserver-configuration-for-using-mobile-app/7414/2
* https://forum.yunohost.org/t/rendre-public-freshrss/7406

I agree with lnoferin in the forum that it could be related to Raspberry Pi.

**Comment by ashwinvis** [comment]
```sh
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Raspbian
Description:	Raspbian GNU/Linux 9.11 (stretch)
Release:	9.11
Codename:	stretch
```

**Comment by lapineige** [comment]
>   I agree with lnoferin in the forum that it could be related to Raspberry Pi.

In what way ?

**Comment by ashwinvis** [comment]
That despite all the fixes which seem to work for others, it does not fix it for use Raspberry Pi users.

**Comment by Alkarex** [comment]
I come from FreshRSS and I am not so familiar with YunoHost, but please try to enable the following two lines

https://github.com/FreshRSS/FreshRSS/blob/b20fddc330932c55945534f9a219d81aefaf62c0/p/api/greader.php#L900-L901

And after trying to log in through the API, provide the resulting log : `./FreshRSS/data/users/_/log_api.txt`

**Comment by lapineige** [comment]
> That despite all the fixes which seem to work for others, it does not fix it for use Raspberry Pi users.

I tried it, indeed, it doesn't work on my Raspberry Pi.

**Comment by ashwinvis** [comment]
@Alkarex I tried that and tried to access the API using FeedReader (following the [instructions](https://freshrss.github.io/FreshRSS/en/users/06_Mobile_access.html)). No `log_api.txt` is generated and FeedReader also complains:

> No connection to the server...

Some config files for reference:

#### /var/www/freshrss/data/config.php
```php
<?php
 return array (
  'salt' => 'xxxxxx',
  'db' =>
  array (
    'type' => 'mysql',
    'host' => 'localhost',
    'user' => 'freshrss',
    'password' => 'xxxxxx',
    'base' => 'freshrss',
    'prefix' => 'freshrss_',
    'pdo_options' =>
    array (
    ),
  ),
  'environment' => 'production',
  'base_url' => 'https://xxxxxx.noho.st/rss',
  'title' => 'FreshRSS',
  'default_user' => 'avmo',
  'auth_type' => 'http_auth',
  'api_enabled' => true,
  'pubsubhubbub_enabled' => true,
);
```

#### /etc/nginx/conf.d/xxxxxx.noho.st.d/freshrss.conf

```nginx
 rewrite ^/rss$ /rss/ permanent;
location /rss/ {
       alias /var/www/freshrss/p/;
       if ($scheme = http) {
            rewrite ^ https://$server_name$request_uri? permanent;
       }

### Example PHP configuration (remove it if not used)
       index index.html index.php;
       try_files $uri $uri/ /index.php?$args;
       location ~ [^/]\.php(/|$) {
           fastcgi_split_path_info ^(.+?\.php)(/.*)$;
           fastcgi_pass unix:/var/run/php/php7.0-fpm-freshrss.sock;

           fastcgi_index index.php;
           include fastcgi_params;
           fastcgi_param REMOTE_USER $remote_user;
           fastcgi_param PATH_INFO $fastcgi_path_info;
           fastcgi_param SCRIPT_FILENAME $request_filename;
       }
### End of PHP configuration part

  # Include SSOWAT user panel.
  include conf.d/yunohost_panel.conf.inc;
}
```

**Comment by ashwinvis** [comment]
@Alkarex My bad! I was looking at the wrong directory. FeedReader says "offline" now and here is the log you asked:

https://paste.yunohost.org/alalawifuf

**Comment by Alkarex** [comment]
I think the `nginx` config might be wrong (as usual, I would say :-P) My guess is that one of the includes defines HTTP Basic Authentication with a global `auth_basic` rule, but I cannot see any exception for `/api/`. Only the sub-directory `/i/` is supposed to be protected by `auth_basic`, not the root `/` nor other sub-folders such as `/api/`.
You can try to browse to `https://xxxxxx.noho.st/rss/api/` from a blank browser and this should work without having to log in. If a YunoHost login is asked, then the YunoHost config is wrong.
Tthis is not related to running on a Raspberry Pi or not.
Maybe not applicable for YunoHost, but just for the record: those things are automatically set up correctly with Apache :-)

**Comment by lapineige** [comment]
For my part this /app/ folder is not protected by auth_basic

**Comment by Alkarex** [comment]
Hum, it was supposed to be fixed by https://github.com/YunoHost-Apps/freshrss_ynh/pull/61
What version are you running?

**Comment by Alkarex** [comment]
@lapineige Public access, and you observe the problem nevertheless?

**Comment by ashwinvis** [comment]
>  You can try to browse to https://xxxxxx.noho.st/rss/api/ from a blank browser and this should work without having to log in. If a YunoHost login is asked, then the YunoHost config is wrong.

The YunoHost login page appears when I try that. Let me know how to proceed.

Here are the includes:


#### /etc/nginx/conf.d/yunohost_panel.conf.inc
```nginx
# Insert YunoHost button + portal overlay
sub_filter </head> '<script type="text/javascript" src="/ynh_portal.js"></script><link type="text/css" rel="stylesheet" href="/ynh_overlay.css"></link><script type="text/javascript" src="/ynhtheme/custom_portal.js"></script><link type="text/css" rel="stylesheet" href="/ynhtheme/custom_overlay.css"></link></head>';
sub_filter_once on;
# Apply to other mime types than text/html
sub_filter_types application/xhtml+xml;
# Prevent YunoHost panel files from being blocked by specific app rules
location ~ (ynh_portal.js|ynh_overlay.css|ynh_userinfo.json) {
}
```

**Comment by Alkarex** [comment]
@ashwinvis Please delete your previous comment (the one with the log) and change your password :-)

**Comment by ashwinvis** [comment]
Of course, I was horrified to see that. Changed it already.

**Comment by Alkarex** [comment]
@ashwinvis I am not familiar enough with YunoHost to guide you. Ping @plopoyop
Please check first that your YunoHost and FreshRSS packages are up to date.
@lapineige Your problem is different then. Please send me the URL of your API end-point if you can.

**Comment by ashwinvis** [comment]
@plopoyop FreshRSS is up-to-date with `1.14.2~ynh1`. There are some php packages with pending updates, such as `
php7.0-gmp (from 7.0.33-0+deb9u3 to 7.0.33-0+deb9u5)
`. I will try that and see if it helps.

**Comment by plopoyop** [comment]
Looks like the unprotected_regex is not working.
for now, you can login in ssh and type : 
`yunohost app setting freshrss unprotected_uris -v "/api/"`
It should open the /api url 
So, you can go back on /rss/api and check if it working when not logged in .

**Comment by ashwinvis** [comment]
@plopoyop I tried it, but it still redirects to the YunoHost SSO page.

**Comment by plopoyop** [comment]
@ashwinvis : ok, lets try this one that just worked on my test instance: 
```
yunohost app setting freshrss unprotected_uris -v "/api/,/scripts/api.js"
yunohost app setting freshrss unprotected_regex -d
```

**Comment by Alkarex** [comment]
@plopoyop For the record, the only path to protect is `/i/` while the other ones such as `/api/` , `/scripts/` , `/themes/` , `/*.*` are better without protection

**Comment by lapineige** [comment]
> /scripts/ , /themes/ , /*.* are better without protection

Why ?

**Comment by Alkarex** [comment]
So that they can be accessed when landing on `/` (e.g. the landing page with FreshRSS logo, robots.txt, favicon.ico)

**Comment by Alkarex** [comment]
But for the landing page, it depends a bit of what one wants with YunoHost, and it may be better to show a YunoHost page, I am not sure. At least on FreshRSS side, only `/i/` needs password access

**Comment by plopoyop** [comment]
Freshrss landing page is not accessible without being logged in.
Since we use freshrss http auth, everything except the api is accessible only when logged

**Comment by ashwinvis** [comment]
So the good news is I get

> Google Reader API configuration test:
>    ✔️ PASS

The bad news is FeedReader still fails to access the api . Here is the end of `log_api.txt`:

```php
[Mon, 21 Oct 2019 23:07:06 +0200] [warning] --- badRequest() Array
(
    [date] => 2019-10-21T23:07:06+02:00
    [headers] => Array
        (
            [Connection] => Keep-Alive
            [User-Agent] => FeedReader 2.10.0
            [Accept-Encoding] => gzip, deflate
            [Host] => ana.noho.st
        )

    [_SERVER] => Array
        (
            [USER] => freshrss
            [HOME] => /home/freshrss
            [HTTP_CONNECTION] => Keep-Alive
            [HTTP_USER_AGENT] => FeedReader 2.10.0
            [HTTP_ACCEPT_ENCODING] => gzip, deflate
            [HTTP_HOST] => xxxxxx.noho.st
            [SCRIPT_FILENAME] => /var/www/freshrss/p/api/greader.php
            [PATH_INFO] =>
            [REMOTE_USER] =>
            [REDIRECT_STATUS] => 200
            [SERVER_NAME] => xxxxxx.noho.st
            [SERVER_PORT] => 443
            [SERVER_ADDR] => 192.168.1.100
            [REMOTE_PORT] => 40098
            [REMOTE_ADDR] => ##.##.##.##
            [SERVER_SOFTWARE] => nginx/1.10.3
            [GATEWAY_INTERFACE] => CGI/1.1
            [HTTPS] => on
            [REQUEST_SCHEME] => https
            [SERVER_PROTOCOL] => HTTP/1.1
            [DOCUMENT_ROOT] => /var/www/freshrss/p/
            [DOCUMENT_URI] => /rss/api/greader.php
            [REQUEST_URI] => /rss/api/greader.php
            [SCRIPT_NAME] => /rss/api/greader.php
            [CONTENT_LENGTH] =>
            [CONTENT_TYPE] =>
            [REQUEST_METHOD] => HEAD
            [QUERY_STRING] =>
            [FCGI_ROLE] => RESPONDER
            [PHP_SELF] => /rss/api/greader.php
            [REQUEST_TIME_FLOAT] => 1571692026.2478
            [REQUEST_TIME] => 1571692026
        )

    [_GET] => Array
        (
        )

    [_POST] => Array
        (
        )

    [_COOKIE] => Array
        (
        )

    [INPUT] =>
)
```

I can open another issue if needed.

**Comment by plopoyop** [comment]
No need to open another issue, lets stick on this one.
I'll test FeedReader on my instance to check

**Comment by Alkarex** [comment]
The rest of the URL that comes after `/api/greader.php` seems to be missing

**Comment by plopoyop** [comment]
On my server : 
```
xx.xx.xx.xx - - [24/Oct/2019:13:55:14 +0200] "HEAD /rss/ HTTP/1.1" 302 0 "-" "FeedReader 2.10.0"
xx.xx.xx.xx - - [24/Oct/2019:13:55:14 +0200] "HEAD /yunohost/sso/?r=aHR0cHM6Ly9odWJlcnQuZmFtaWx5L3Jzcy8= HTTP/1.1" 404 0 "-" "FeedReader 2.10.0"
```

It looks like FeedReader is making a HEAD request on `/api/` and get  SSOed
There is a possibilty to add / to unprotected url, so, when you are not logged in your web browser you get the 403 error from freshrss. But when you are logged in, you get your rss. And so, feedreader should work (did not have the time to test it myself).
I won't recommand this solution. You better check with feedreader if the HEAD request could be done on `/rss/api/`

**Comment by lapineige** [comment]
@plopoyop does that means all clients need to be changed ?

**Comment by plopoyop** [comment]
@lapineige only this client.
The purpose of a "HEAD" request is to know the headers that will be send by the server for a request before sending this request.
FeedReader is making a HEAD request on `/rss/` before sending the request on `/rss/api/greader.php`
Therefore, the HEAD request is intercepted by the sso and the client won't continue.
I am using EasyRss client which works perfectly.
FeedReader must send the HEAD request on `/rss/api/greader.php` and that will work.

**Comment by Alkarex** [comment]
@plopoyop Let me know if something can be done on FreshRSS side to help.

**Comment by Fipaddict** [comment]
Hi, 
I don't know if that can help to find a solution but just to inform you that I have now this problem (apparently since my last YNH & FreshRSS update...)

YNH -> 3.7.0.12
FresshRSS -> 1.15.3~ynh1

**Comment by anmol26s** [comment]
1. Go to `https://domain.tld/rss/i/?c=user&a=profile` and fill the API password. 
2. Run these commands https://github.com/YunoHost-Apps/freshrss_ynh/issues/80#issuecomment-544699687
3. Login to FeedReader with `https://domain.tld/rss/api/greader.php` as **URL** , your **username** of Yunohost and **password** given in step 1.

**Comment by lapineige** [comment]
>Go to https://domain.tld/rss/i/?c=user&a=profile and fill the API password.

I did that, with a different password than Yunohost one, because I thought it was not the same password as FreshRSS account, but dedicated to the API only… now I can't login anymore :(

**Comment by anmol26s** [comment]
> I did that, with a different password than Yunohost one, because I thought it was not the same password as FreshRSS account, but dedicated to the API only… now I can't login anymore :(

You can put any password in API password. It should work. Your login has nothing to do with API password.

**Comment by lapineige** [comment]
I have an error page with:
> Error 403 - Forbidden
Vous n’avez pas le droit d’accéder à cette page ! [HTTP Remote-User=]

I'm wondering… maybe when I put a password in the API password field, and nothing in the user password field, it changed also the user password? :thinking:

@Alkarex is there a way to change the password from the command line ?

**Comment by plopoyop** [comment]
Did you put freshrss into visitor group permission ?

**Comment by lapineige** [comment]
I don't think I did anything like this 😅

But I used your command from that comment https://github.com/YunoHost-Apps/freshrss_ynh/issues/80#issuecomment-544699687

**Comment by plopoyop** [comment]
that's weird, this error is when you are not connected to ynh and there is no "protection" on freshrss /i/ path.
Do you have this error even if you go on your ynh instance to connect and go to freshrssh via the tile on the ynh home page ?

**Comment by lapineige** [comment]
Oh, good point… I don't have it 🤔
But then when I click on an article, I'm redirected to this error page.
And… yes I can access FreshRSS (this error page in fact) without login in Yunohost :( (it should be behind the SSO)

**Comment by lapineige** [comment]
I made a few more test.

If I connect thought Yunohost interface, I can access FreshRSS.
Then I can go to the settings for instance, and even change API password (and that don't fix https://github.com/YunoHost-Apps/freshrss_ynh/issues/80#issuecomment-616101238).
But if I click on an article (it should mark it as read), I'm logged out and I have the error page. "Normally" I should be redirected to the SSO, as there should be no public access to my FreshRSS (even if it can't login).
So it just seems that I excluded the login page from the SSO. And my log-out issue is still there #68.
The mobile app still don't work.

**Comment by lapineige** [comment]
>     1. Go to `https://domain.tld/rss/i/?c=user&a=profile` and fill the API password.
> 
>     2. Run these commands [#80 (comment)](https://github.com/YunoHost-Apps/freshrss_ynh/issues/80#issuecomment-544699687)
> 
>     3. Login to FeedReader with `https://domain.tld/rss/api/greader.php` as **URL** , your **username** of Yunohost and **password** given in step 1.

That only can work if the client allows to set the whole connection URL.
But FreshRSS app from f-droid, even if it uses the greader API, won't allow us to do that. It autocomplete the `/ap/greader/php`. #69
I've reported that issue.

**Comment by lapineige** [comment]
Maybe we should test this again after #127

**Comment by yalh76** [comment]
should be fixed by now
